Container -> Menu

Juuxel opened the pull request:
Replaces `Container` with `Menu`, as seen in various registry and logging strings. Closes #386 and #889.

Prospector commented on mappings/net/minecraft/client/gui/screen/ingame/AbstractMenuScreen.mapping:
```diff
@@ -1,9 +1,9 @@
-CLASS net/minecraft/class_465 net/minecraft/client/gui/screen/ingame/AbstractContainerScreen
+CLASS net/minecraft/class_465 net/minecraft/client/gui/screen/ingame/AbstractMenuScreen
```
Let's drop the Abstract here

Juuxel replied:
Done.

modmuss50 commented:
Why `Menu` ? A few days ago we were talking about names such as `ScreenController`. `ScreenSate` `ScreenHandler`.

A menu to me seems something is closer to a UI and not something that partly handles the syncing of it?

Juuxel commented:
`Menu` is the name used by Mojang in the menu/container/controller registry. I think it's important to use Mojang names for clarity; in addition to the registry name and strings in methods, there are unobfuscated methods like `createMenu` that don't really make sense if the class is something else than `Menu`.

`Menu` is also a much better name than `Container`, as well as simple and short. Names like `ScreenController` are long and don't work that well in some interface names, like `NameableScreenControllerProvider`.

> A menu to me seems something is closer to a UI

Menus also handle some parts of the UI, mainly slots.

Pyrofab commented:
I would take #351 into consideration before going with the Mojang name. I think `Menu` is an unintuitive name for a class that acts as intermediate between an Inventory (or another backend) and a Screen.
As was noted in that issue, the class is rather reminiscent of a Controller, using the MVC pattern as a reference. However, as you noted, `ScreenController` becomes quite long in compound names. My suggestion would be to use `ScreenController` for the base class, but only using `Controller` in other class names. That way, the purpose can always be deducted easily from the context without relying on painfully long names.
A meaningless Mojang name is certainly a step up from a meaningless stolen name, but I think we can do more helpful than that.

liach commented:
To me, it feels like the receiver/controller chips in a remote. Maybe we can consider the name from this viewpoint.

Prospector commented:
I think the arguments for Menu are pretty solid, not to mention the unobfuscated method `createMenu`... it's just blatantly confusing to name it anything else. We have MULTIPLE places in the code that reference it as a menu. Let's just make it simple and adopt the name. I think anything else would be rather silly.

stefan134 commented:
Yes, I agree with @Prospector. On top of having references in unobfuscated names, there are also strings (exception messages, and even the ID of the registry!) using the word "menu". 

Using a different name would be the same as using two different names. A mix of a good name and a bad name is worse than a bad name.

Why aren't we also arguing about "entity". The word entity is so vague that anything could be an entity, yet we still use it despite having better alternatives.

Also, what about "block entity"? Block entities are in no way related to entities. Wouldn't "block data" be a better description of what it does?

liach commented:
This pr has no problem as I see. Should we go to #386 for whether to rename or create a new ticket for a new vote?

sfPlayer1 commented:
https://en.wikipedia.org/wiki/Menu_(computing)

There is nothing menu-like about this class, even less so than in the Gui class hierarchy.

liach commented:
Please change base branch to 19w42a.

liach commented:
Also consider renaming all gui packets to menu packets.

liach commented:
Please change base branch to 19w44a.

Sollace commented:
> 
> 
> To me, it feels like the receiver/controller chips in a remote. Maybe we can consider the name from this viewpoint.

`RemoteInventory`, `RemoteScreen`?

Sollace commented:
`SpookyIntactionsAtADistance`

Sollace commented:
@Juuxel This history is getting pretty messy. Would you consider rebasing?

Juuxel commented:
I squashed the commits, should be cleaner now with all the merges gone.

Juuxel commented:
Please switch the base branch to 1.15-pre4 and resolve the conflicts.

liach commented:
Wait, juuz, are you talking to yourself

Juuxel commented:
Updated this to 1.15-pre6.

Sollace reviewed (commented):
Best commit.

Juuxel commented:
Oh, 542bf10? VSCode didn't pick them up as files with conflicts so I didn't notice them when I merged.

Juuxel commented:
Updated to 1.15-pre7.

liach commented:
will we keep this open? this pr is overwhelmingly downvoted

Juuxel commented:
There are also [Runemoro's statistics](https://github.com/FabricMC/yarn/pull/846#issuecomment-540032393), which are a bit more comprehensive than PR votes.

modmuss50 commented:
I think the issue still stands, im not sure Menu is the correct solution to it though

Juuxel commented:
Well, I'd be happy to see someone else's `Container -> X` PR as well, but so far this has been the only one (in addition to #566).

Prospector commented:
Can we just get Menu in because it's the obvious choice so we can fix the horrible Inventory -> Container thing because we can all unanimously agree Container is incorrect. And then we can open further discussions about it we want to rename Menu.

modmuss50 commented:
Im not sure it is the `obvious choice`, 3/4 of people have thumbs down it on this issue.

It doesnt make a lot of sense, as a menu is somehting that a user interfacts with. `GuiState` seemed to be a popular option on the discord.

Prospector commented:
I say `obvious choice` because it is literally the name vanilla calls it so it is literally a correct name. IMO if anyone is unhappy with this and has a better name, they should open their own PR. As long as we change this name from Container I will be happy for now.

Juuxel commented:
Should I do something about the conflicting `ContainerProvider`s (inherited as two different `MenuProvider`s) in this PR? One is a more general superinterface of `NameableMenuProvider` (for blocks and BEs), the other a general interface for `MenuScreen`-like screens.

Juuxel committed:
Container -> Menu

liach commented:
I hereby support the name `Menu`. It indeed offers a list of item (see [definition of menu on dictionary.com](https://www.dictionary.com/browse/menu) `any list or set of items, activities, etc., from which to choose: `) as its main functionality.

Juuxel committed:
Javadoc

liach commented on mappings/net/minecraft/menu/Menu.mapping:
```diff
@@ -1,4 +1,17 @@
 CLASS net/minecraft/class_1703 net/minecraft/menu/Menu
+	COMMENT A menu is an object used for synchronizing GUIs between the server and the client.
```
Not really synchronizing GUIs but more exactly, synchronizing screens that have item lists (menus).
I'd say:
`A menu is an item list, synchronized between the server and the client to allow updating the client to modify it.`

liach commented on mappings/net/minecraft/menu/Menu.mapping:
```diff
@@ -1,4 +1,17 @@
+	COMMENT that are used for syncing integers (e.g. progress bars) from the server to the client.
+	COMMENT
+	COMMENT <p>The {@link #syncId} is shared between the two sides. It is used to verify that a player
+	COMMENT has a menu open so that they can move items, for example.
```
The sync id is just an identifier for the menu so as to avoid desyncs, isn't it

Juuxel replied:
IIRC it's also used to check if a player has a menu open when there's a slot update packet etc.

liach commented on mappings/net/minecraft/menu/Menu.mapping:
```diff
@@ -1,4 +1,17 @@
+	COMMENT <p>The {@link #syncId} is shared between the two sides. It is used to verify that a player
+	COMMENT has a menu open so that they can move items, for example.
+	COMMENT
+	COMMENT <p>On the client, menus are visualized by a {@link net.minecraft.client.gui.screen.ingame.MenuScreen}.
```
maybe mention it has a reference to a client menu?

Juuxel replied:
Sure.

liach commented on mappings/net/minecraft/menu/GenericMenu.mapping:
```diff
@@ -1,8 +1,20 @@
 CLASS net/minecraft/class_1707 net/minecraft/menu/GenericMenu
+	COMMENT A generic 9xN menu with a customizable amount of rows.
```
`A generic menu with 9 columns and a customizable amount of rows.`

liach commented on mappings/net/minecraft/menu/GenericMenu.mapping:
```diff
@@ -1,8 +1,20 @@
+	COMMENT A generic 9xN menu with a customizable amount of rows.
+	COMMENT
+	COMMENT <p>The {@code createGeneric9xN} methods should usually be used only on the client,
+	COMMENT as they have an empty inventory by default.
```
Maybe mention that they can be updated by s2c packets and have c2s packets to response?

Juuxel replied:
I think that that would work better in the doc on `Menu`.

Juuxel replied:
I think the comment I have on `Menu` already describes the packet behavior (well, maybe not c2s responses but that's an impl detail).

liach replied:
worth nothing that server-only fab mods can send a simple generic menu to client and the client will have no trouble interacting with the menu

Juuxel committed:
Small jd changes

Juuxel committed:
More property jd

Juuxel committed:
Networking stuff

Juuxel committed:
ConfirmGuiAction*Packet tweaks

liach commented on mappings/net/minecraft/network/listener/ServerPlayPacketListener.mapping:
```diff
@@ -17,7 +17,7 @@ CLASS net/minecraft/class_2792 net/minecraft/network/listener/ServerPlayPacketLi
 	METHOD method_12053 onBookUpdate (Lnet/minecraft/class_2820;)V
 		ARG 1 packet
-	METHOD method_12054 onGuiClose (Lnet/minecraft/class_2815;)V
+	METHOD method_12054 onCloseMenu (Lnet/minecraft/class_2815;)V
```
```suggestion
	METHOD method_12054 onMenuClose (Lnet/minecraft/class_2815;)V
```
imo the packet doesnt' need renaming but this method name is more congruent with other methods.

Juuxel committed:
Update mappings/net/minecraft/network/listener/ServerPlayPacketListener.mapping

Co-Authored-By: liach <7806504+liach@users.noreply.github.com>

liach commented on mappings/net/minecraft/menu/Menu.mapping:
```diff
@@ -1,4 +1,19 @@
-CLASS net/minecraft/class_1703 net/minecraft/container/Container
+CLASS net/minecraft/class_1703 net/minecraft/menu/Menu
+	COMMENT A menu is a list of items and properties, synchronized between the server and the client.
```
To address @sfPlayer1's concern at https://github.com/FabricMC/yarn/pull/846#issuecomment-540034825, we should point out that this is not a menu in computing sense at all.

Juuxel committed:
Merge branch '20w10a' into menu

Juuxel commented:
This should be up to date for 20w10a now.

liach commented on mappings/net/minecraft/entity/vehicle/StorageMinecartEntity.mapping:
```diff
@@ -8,11 +8,7 @@ CLASS net/minecraft/class_1693 net/minecraft/entity/vehicle/StorageMinecartEntit
-	METHOD createMenu (ILnet/minecraft/class_1661;Lnet/minecraft/class_1657;)Lnet/minecraft/class_1703;
-		ARG 1 syncId
-	METHOD method_17357 createMenu (ILnet/minecraft/class_1661;)Lnet/minecraft/class_1703;
+	METHOD method_17357 getMenu (ILnet/minecraft/class_1661;)Lnet/minecraft/class_1703;
```
so `getMenu` or `createMenu`?

Juuxel replied:
It should be `createMenu`, I think. I'm pretty sure there was a `getContainer` there when I was merging.

Juuxel committed:
StorageMinecartEntity.getMenu -> createMenu

Also mapped checkLootInteraction(), which is exactly like the one in
LootableContainerBlockEntity.

Juuxel committed:
Merge branch '20w10a' into menu

Earthcomputer commented:
I suppose this can be closed now in favour of #1106?

liach commented:
rip our good ol' pr
